### PR TITLE
[SDK] Feature: Ox Transaction Envelopes

### PR DESCRIPTION
**Pull request by gregfromstl**
CNCT-2303

Woah. This is sick.

---

## PR-Codex overview
This PR focuses on updating the `thirdweb` package to improve transaction serialization by using the `ox` library, along with refactoring and renaming some files and functions for consistency.

### Detailed summary
- Renamed `coinbaseWebSDK.js` to `coinbase-web.js` across multiple files.
- Updated imports for `parseTypedData` to use `parse-typed-data.js`.
- Introduced `SerializableTransaction` type in various files.
- Added `SignableMessage` type definition.
- Implemented `parseTypedData` function to handle hex `chainId`.
- Updated `sign`, `signTransaction`, and `signTypedData` functions to use `ox` library for signing.
- Refactored tests to align with new implementations and types.
- Implemented error handling in keyless transaction methods.
- Enhanced Coinbase wallet functionalities with mobile provider support.

> The following files were skipped due to too many changes: `packages/thirdweb/src/adapters/ethers5.test.ts`, `packages/thirdweb/src/utils/hashing/hashTypedData.ts`, `packages/thirdweb/src/wallets/coinbase/coinbase-web.test.ts`, `packages/thirdweb/src/transaction/serialize-transaction.ts`, `packages/thirdweb/src/transaction/serialize-transaction.test.ts`, `pnpm-lock.yaml`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/thirdweb/src/utils/signatures/sign-typed-data.test.ts [hunk 1]
```diff
@@ -83,21 +83,6 @@ test("domain: chainId", async () => {
   );
 });
 
-test("domain: chainId hex", async () => {
```
Comment (jnsdls) [inline packages/thirdweb/src/utils/signatures/sign-typed-data.test.ts:86]: why the removed test?

File: packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts [hunk 2]
```diff
@@ -42,15 +42,19 @@ const MAGIC_VALUE = "0x1626ba7e";
...
+  typedData extends ox__TypedData.TypedData | Record<string, unknown>,
   primaryType extends keyof typedData | "EIP712Domain" = keyof typedData,
 >(options: CheckContractWalletSignTypedDataOptions<typedData, primaryType>) {
   if (!isHex(options.signature)) {
     throw new Error("The signature must be a valid hex string.");
   }
   const result = await isValidSignature({
     contract: options.contract,
-    hash: hashTypedData(options.data),
+    hash: ox__TypedData.hashStruct({
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts:53]: ugh types clashed here?

File: .changeset/little-kids-wash.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"thirdweb": patch
```
Comment (joaquim-verges) [inline .changeset/little-kids-wash.md:2]: would make this a minor - lots of underlying changes

File: packages/thirdweb/src/transaction/actions/sign-transaction.ts [hunk 4]
```diff
@@ -1,11 +1,13 @@
-import type { TransactionSerializable } from "viem";
+import * as ox__Hash from "ox/Hash";
```
Comment (joaquim-verges) [inline packages/thirdweb/src/transaction/actions/sign-transaction.ts:1]: curious why you dont just `import { Hash } from 'ox'` ?

File: packages/thirdweb/src/transaction/serialize-transaction.test.ts [hunk 5]
```diff
@@ -24,29 +16,31 @@ const BASE_TRANSACTION = {
...
       transaction: baseEip1559,
     });
-    assertType<TransactionSerializedEIP1559>(serialized);
     expect(serialized).toEqual(
       "0x02ef0182031184773594008477359400809470997970c51812dc3a010c7d01b50e0d17dc79c8880de0b6b3a764000080c0",
     );
 
-    const tx = parseTransaction(serialized);
+    const tx = ox__TransactionEnvelopeEip1559.deserialize(
+      serialized as ox__TransactionEnvelopeEip1559.Serialized,
```
Comment (joaquim-verges) [inline packages/thirdweb/src/transaction/serialize-transaction.test.ts:38]: will these cast go away once we migrate everything?

File: packages/thirdweb/src/transaction/actions/sign-transaction.ts (hunk 4 above)
Comment (gregfromstl) [reply packages/thirdweb/src/transaction/actions/sign-transaction.ts:1]: cjs won't treeshake properly if I do, have to import it this way

File: packages/thirdweb/src/transaction/serialize-transaction.test.ts (hunk 5 above)
Comment (gregfromstl) [reply packages/thirdweb/src/transaction/serialize-transaction.test.ts:38]: These are just the tests being extra picky because I'm allowing legacy transactions

File: packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts (hunk 2 above)
Comment (gregfromstl) [reply packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts:53]: No I'm just trying to use ox as much as possible in the underlying functions, so we don't need to wrap and/or rewrite internal utils we're not planning to export

File: packages/thirdweb/src/utils/signatures/sign-typed-data.test.ts (hunk 1 above)
Comment (gregfromstl) [reply packages/thirdweb/src/utils/signatures/sign-typed-data.test.ts:86]: This was an edge case we had for wallet connect connections which I believe is resolved now, you shouldn't expect chain ID to come in as a hex

File: packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts (hunk 2 above)
Comment (joaquim-verges) [reply packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts:53]: i meant the casts underneath this

File: packages/thirdweb/src/transaction/actions/sign-transaction.ts (hunk 4 above)
Comment (joaquim-verges) [reply packages/thirdweb/src/transaction/actions/sign-transaction.ts:1]: really? ugh

**Review by joaquim-verges — APPROVED** [review]

File: packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts (hunk 2 above)
Comment (gregfromstl) [reply packages/thirdweb/src/extensions/erc1271/checkContractWalletSignedTypedData.ts:53]: Oh yes they do

**Comment by gregfromstl** [comment]
### Merge activity

* **Dec 6, 5:21 PM EST**: The merge label 'merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=thirdweb-dev&repo=js) once it meets the requirements.
